File: peeking-iterator/shaun-gy-flag.cpp
// Below is the interface for Iterator, which is already defined for you.
// **DO NOT** modify the interface for Iterator.
//grandyang idea using a flag and value copy
//shaun read and understand
//runtime 4ms, faster than 83%, mem less than 100%
class Iterator {
    struct Data;
    Data* data;
public:
    Iterator(const vector<int>& nums);
    Iterator(const Iterator& iter);
    virtual ~Iterator();
    // Returns the next element in the iteration.
    int next();
    // Returns true if the iteration has more elements.
    bool hasNext() const;
};

class PeekingIterator : public Iterator {
public:
    PeekingIterator(const vector<int>& nums) : Iterator(nums) {
        _flag = false;
    }

    int peek() {
        if (!_flag) _value = Iterator::next();
        _flag = true;
        return _value;
    }

    int next() {
        if (!_flag) return Iterator::next();
        _flag = false;
        return _value;
    }

    bool hasNext() const {
        return _flag || Iterator::hasNext();
    }

private:
    int _value;
    bool _flag;
};

File: peeking-iterator/shaun-gy-flag.java
// Java Iterator interface reference:
// https://docs.oracle.com/javase/8/docs/api/java/util/Iterator.html
//shaun code from gy's 1 register idea, accepted after 1 bug fix, always think if the register is used or not
//runtime 45ms, faster than 96%, mem less than 100%
class PeekingIterator implements Iterator<Integer> {
    private boolean flag = false;
    private int copy;
    private Iterator<Integer> it;
	public PeekingIterator(Iterator<Integer> iterator) {
	    // initialize any member here.
	    it = iterator;
	}

    // Returns the next element in the iteration without advancing the iterator.
	public Integer peek() {
        if (!flag) {
            copy = it.next();
            flag = true;  //registered used
            return copy;
        }
        //flag true, register used
        return copy;
	}

	// hasNext() and next() should behave the same as in the Iterator interface.
	// Override them if needed.
	@Override
	public Integer next() {
	    if (flag) {
            flag = false;
            return copy;
        }
        return it.next();
	}

	@Override
	public boolean hasNext() {
	    return flag || it.hasNext();
	}
}
